### ISSUE TITLE
known_hosts module not working ?

### ISSUE THREAD

**Issue by mmallejac** [ansible/ansible-modules-extras#664]
##### Issue Type:

Bug Report
##### Component Name:

known_hosts
##### Ansible Version:

ansible --version
ansible 1.9.2
##### Ansible Configuration:

N/A
##### Environment:

apt-add-repository ppa:ansible/ansible
aptitude update
aptitude install ansible python-pip 

on Ubuntu 14.04.2 LTS
##### Summary:

This is a copy/confirmation of https://github.com/ansible/ansible/issues/11222
Trying for instance to add his own public key to itself, so ssh root@localhost would work, maybe not using the module correctly ?
##### Steps To Reproduce:

```
  - name: add itself to known_hosts list
    known_hosts: host={{ inventory_hostname }} key="{{ lookup('file', dataDir + '/' + inventory_hostname + '/id_rsa.pub') }}"
```
##### Expected Results:

on target host, ssh localhost should work without asking add fingerprint
##### Actual Results:

```
failed: [myHost] => {"cmd": "/usr/bin/ssh-keygen -F myHost -f /tmp/tmpE7kC4N", "failed": true, "rc": 1}
stderr: line 1 invalid key: ssh-rsa...
/tmp/tmpE7kC4N is not a valid known_hosts file.

msg: line 1 invalid key: ssh-rsa...
/tmp/tmpE7kC4N is not a valid known_hosts file.
```

**Comment by sivel** [comment]
I think this may possibly be a misunderstanding.

The error shown is telling you that the key is not a proper SSH host key.

Looking at your lookup, I am assuming that the key you are giving it is a ssh public key, that would normally be used with `authorized_keys` as opposed to `known_hosts`.

The `known_hosts` module, is meant to manage `~/.ssh/known_hosts`, whereas the `authorized_keys` module would manage `~/.ssh/authorized_keys`.

ansible/ansible#11222 is a different issue, in that the key string was missing the key type.

**Comment by costela** [comment]
This might be partly a misunderstanding, but the issue seems real enough. Host keys and user pubkeys have TTBOMK the same format, which means this should work, regardless of semantics.

My first guess, without having looked at the code yet, is that the module checks the syntax of the generated known_hosts file before moving it to its final location. This check fails if the added line uses the pubkey format, because the hostname is missing as first token in the line. 
If you manually prepend the hostname to the pubkey, it seems to work*. Maybe the module could intelligently add the hostname to the pubkey if it detects the format? (since the "host"/"name" parameter is required anyway)

\* currently, if I manually prepend the hostname, I then get the problem that the module always returns changed=true (which is admittedly less of an issue).

**Comment by letompouce** [comment]
I confirm the module works, more or less :)

@costela : Probably the doc should have a Note about it.
There is already some check there. Consider this snippet:

```
- name: tell the host about our servers it might want to ssh to
  known_hosts:
    host='example.com'
    path='/etc/ssh/ssh_known_hosts'
    key="and.now.for.something.very.different ssh-rsa AAAAAxxxxxxxx"
```

You'll get an meaningless error message:

```
failed: [thishost] => {"cmd": "/usr/bin/ssh-keygen -F example.com -f /tmp/tmpc18Ubq", "failed": true, "rc": 1}
```

BTW, the task always returns `changed=true` because the module overwrites the file every time. See #691 about this.

**Comment by truptimali** [comment]
I am also facing the same issue here. 
failed: [node1] => {"cmd": "/usr/bin/ssh-keygen -F xx.xx.xx.xx -f /tmp/tmpaZMs0F", "failed": true, "rc": 1}
stderr: line 1 invalid key: ssh-rsa...
/tmp/tmpaZMs0F is not a valid known_hosts file.

my code is <code>
- name: manage known hosts 
  known_hosts: path=/home/{{ user }}/.ssh/known_hosts  host={{ localhost }} key=~/.ssh/id_rsa.pub
  </code>

**Comment by letompouce** [comment]
@truptimali :
The `key=` value must be the key string itself. See my example.

**Comment by truptimali** [comment]
@letompouce . Thanks for the response. I was trying out something and pasted the same thing here by mistake. But my original code looks like this -<code>
- name: manage known hosts 
  known_hosts: path=/home/{{ user }}/.ssh/known_hosts  host={{ localhost }} key="{{ lookup('file', '~/.ssh/id_rsa.pub') }}"
  </code>

but then I referred to the below link and added hostname in the beginning of the key and it worked for me.

key="{{ localhost }} {{ lookup('file', '~/.ssh/id_rsa.pub') }}"

https://groups.google.com/forum/?utm_medium=email&utm_source=footer#!msg/ansible-project/jGVJuOT5lDk/Zo800XXiot8J
I think the known-hosts module should take care of prepending the hostname specified in the parameters list.
Thanks
Trupti

**Comment by mahemoff** [comment]
_Update - ignore my comment here. @ahes has a simpler solution_

<s>Okay for anyone else who got confused and landed here from search, do something like below.

<s>This example assumes you want to add github as a known host.</s>

<s>1. From target host, ssh to github.com.
2. `cat ~/.ssh/known_hosts` and find the line with github.com
3. Copy that line to your Ansible project as a separate file named 'github_key.txt', e.g. make a one-line file with the content: `github.com ssh-rsa ABChundredsofgarbagechars1211183u9zp==`
4. Add an ansible task like:</s>

<s>```
known_hosts:
  path: /home/player/.ssh/known_hosts
  name: github.com
  key: "{{ lookup('file', ''/path/to/github_key.txt')  }}"</s>

**Comment by marcinhlybin** [comment]
@mahemoff 
It is better to scan for public key:
`ssh-keyscan -T 10 github.com`

or with hashed hosts:
`ssh-keyscan -H -T 10 github.com`

Working task could look like:

```
- name: Add backup servers to known hosts
  tags: ['backup', 'backup-ssh']
  known_hosts:
    host={{ item }}
    path=/root/.ssh/known_hosts    
    key="{{ lookup('file', '{{ inventory_dir }}/ssh/servers/{{ item }}.pub') }}"
  with_items: backup_servers
```

Now the funny thing how I got here. I made a mistake with lookup closing parenthesis in wrong place:
`key="{{ lookup('file'), '{{ inventory_dir }}/ssh/servers/{{ item }}.pub' }}"`

Then content of the key file on the server (in temp path) is `([], u'/Users/ahes/Workspace/Project/ansible-project/ssh/servers/bkp3.host.pub')` and eventually user gets error:

`TASK [backup : Add backup servers to known hosts] ************************
failed: [localhost] => (item=bkp3.host) => {"cmd": "/usr/bin/ssh-keygen -F bkp3.host -f /tmp/tmpY8on_k", "failed": true, "item": "bkp3.host", "msg": "line 1 invalid key: ([],...\r\n/tmp/tmpY8on_k is not a valid known_hosts file.", "rc": 1, "stderr": "line 1 invalid key: ([],...\r\n/tmp/tmpY8on_k is not a valid known_hosts file.\n", "stdout": "", "stdout_lines": []}`

**Comment by RavisMsk** [comment]
@truptimali method worked out well, thanks!

**Comment by ptman** [comment]
Let me still add my own little clarification:

```
known_hosts:
    path: /etc/ssh/ssh_known_hosts
    name: '{{ansible_fqdn}}'
    key: '{{ansible_fqdn}} ecdsa-sha2-nistp256 {{ansible_ssh_host_key_ecdsa_public}}'
delegate_to: '{{target_host}}'
```

**Comment by marcelloromani** [comment]
Disabling hardware acceleration fixed the "show URL on hyperlink hover" issue, but affected page rendering performance pretty badly.

**Comment by marcelloromani** [comment]
I have an issue whereby I keep getting "state": "absent" in the ansible-playbook output, even though I specify state: present in the playbook.

**Comment by costela** [comment]
@marcelloromani your problem doesn't seem to be related to this issue. Please open a separate issue, also including more details like a minimal non-working example.

About this issue: the original misunderstanding may have been solved by #716 . Could we maybe close this?

**Comment by marcelloromani** [comment]
Hmm yes I think so. I'll open a separate issue.

**Comment by ansibot** [comment]
@mcv21, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
@mmallejac ping, this issue is still waiting on your feedback. We will close the issue if you do not respond.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
@mmallejac ping, this issue is still waiting on your feedback. We will close the issue if you do not respond.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
@mmallejac ping, this issue is still waiting on your feedback. We will close the issue if you do not respond.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
@mmallejac closure has been requested for this issue or it has timed out waiting for your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by mjudeikis** [comment]
This might save somebody day or two:

```
- name: Populate known_hosts files on heketi servers
  known_hosts:
    state: present
    path: /etc/ssh/ssh_known_hosts
    name: "{{ item.name | lower }}"
    key: "{{ item.name | lower }},{{ item.ip}} ecdsa-sha2-nistp256 {{item.key}}"
  with_items: "{{ hostkeys.results | json_query ('[].ansible_facts.{ name: ansible_hostname, ip: ansible_default_ipv4.address, key: ansible_ssh_host_key_ecdsa_public }') }}"
```

main area which tripped me was `{{ item.name | lower }}` as guys "here" uses some of the hostnames uppercase and some lower case. So I was seeing intermediate failures
